multinet/__init__.py

```diff
@@ -21,13 +21,48 @@
 sentry_sdk.init(dsn=sentry_dsn, integrations=[FlaskIntegration()])
 
 
-def get_allowed_origins() -> List[str]:
+def get_allowed_origins() -> Set[str]:
     """Read in comma-separated list of allowed origins from environment."""
     allowed_origins = os.getenv("ALLOWED_ORIGINS", default=None)
     if allowed_origins is None:
-        return []
+        return set()
 
-    return [s.strip() for s in allowed_origins.split(",")]
+    return {s.strip() for s in allowed_origins.split(",")}
+
+
+def register_cors_func(app: Flask, with_credentials: bool = False) -> None:
+    """Add the included function to the list of functions to run after a request."""
+    cors_header = "Access-Control-Allow-Origin"
+    cors_credentials_header = "Access-Control-Allow-Credentials"
+
+    def check_origin(res: Response) -> Response:
+        """
+        Match the request origin to the list of wildcard allowed origins.
+
+        If the match succeeds, then CORS headers are added to the response.
+        Otherwise, the response is returned unchanged.
+        """
+        allowed_origins: Set[str] = app.config["ALLOWED_ORIGINS"]
+        normal_allowed_origins = allowed_origins - {"*"}
+
+        origin = request.environ.get("HTTP_ORIGIN")
+        if origin is None or cors_header in res.headers or not allowed_origins:
+            return res
+
+        # Two tests are done, because if both "*" and the actual origin are present,
+        # we want the actual origin to "win", in case of credential use, since
+        # credential use on origin "*" is not allowed.
```

Comment:
Is this part of the spec or a security check we've added? Would this cause issues with accessing data on the vis app deploy previews? I'm asking since you need to pass credentials to get the data out, and we'd likely use the wildcard for the previews

Reply:
It's part of how CORS works, but it applies specifically to the case where you supply `*` as the allowed origin. This shouldn't interfere with what we want to do, which is allow credentials for allowed origins featuring `*` *as a wildcard*; the approach is to use `fnmatch` to say "yes" or "no" to individual domains so it actually has nothing to do with the restriction on `*` in CORS itself.

@AlmightyYakob, please correct me if I got anything wrong.

Reply:
Yes, it's to match the CORS spec, which is that if the allowed origin is exactly `*`, then credential support is not allowed. Since we can specify multiple allowed origins, this check is to make sure that if we supply a pattern which matches the origin more strictly than `*`, that it is preferred.